File: 1520-maximum-number-of-non-overlapping-substrings/1520-maximum-number-of-non-overlapping-substrings.py
class Solution:
    def maxNumOfSubstrings(self, s: str) -> List[str]:
        start = [inf] * 26
        end = [-1] * 26
        ans = []
        n, a, right = len(s), ord('a'), -1
        for i in range(n):
            idx = ord(s[i]) - a
            start[idx] = min(start[idx], i)
            end[idx] = i
            
        for i in range(n):
            idx = ord(s[i]) - a
            if start[idx] != i:
                continue
            new_right = self.checkSubstring(i, end[idx], s, start, end)
            if new_right != -1:
                if i > right:
                    ans.append('')
                right = new_right
                ans[-1] = s[i:right+1]
        return ans
    
    
    def checkSubstring(self, l, r, s, start, end):
        j, right = l, r
        while j <= right:
            idx = ord(s[j]) - ord('a')
            if start[idx] < l:
                return -1
            right = max(right, end[idx])
            j += 1
        return right
      

File: 1520-maximum-number-of-non-overlapping-substrings/1520-maximum-number-of-non-overlapping-substrings.js
/**
 * @param {string} s
 * @return {string[]}
 */
var maxNumOfSubstrings = function(s) {
  const start = new Array(26).fill(-1);
  const end = new Array(26).fill(-1);
  const n = s.length;
  const ans = [];
  let right = -1;
  
  for (let i = 0; i < n; i++) {
    const idx = charAt(s[i]);
    if (start[idx] == -1) start[idx] = i;
    end[idx] = i;
  }
  
  for (let i = 0; i < n; i++) {
    const idx = charAt(s[i]);
    if (start[idx] !== i) continue;
    
    const newRight = checkSubstringRight(i, s, start, end);
    if (newRight === -1) continue;
    
    if (i > right) ans.push('');
    
    right = newRight;
    ans[ans.length - 1] = s.slice(i, right + 1);
  }
  
  return ans;
};
  
const charAt = (ch) => ch.charCodeAt() - 'a'.charCodeAt(); 

const checkSubstringRight = (l, s, start, end) => {
  let right = end[charAt(s[l])];
  
  for (let i = l; i <= right; i++) {
    let idx = charAt(s[i]);
    if (start[idx] < l) return -1;
    
    right = Math.max(right, end[idx]);
  }
  
  return right;
}